service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/BlobCopier.java

```diff
@@ -0,0 +1,110 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure;
+
+import static java.util.stream.Collectors.groupingBy;
+
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storage.ControlledAzureStorageResource;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storageContainer.ControlledAzureStorageContainerResource;
+import com.azure.core.util.polling.PollResponse;
+import com.azure.core.util.polling.SyncPoller;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.models.BlobCopyInfo;
+import com.azure.storage.blob.models.BlobItem;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BlobCopier {
+  private final Logger logger = LoggerFactory.getLogger(BlobCopier.class);
+  private final AzureStorageAccessService storageAccessService;
+  private final AuthenticatedUserRequest userRequest;
+
+  private static final Duration MAX_POLL_TIMEOUT = Duration.ofMinutes(60);
+
+  public BlobCopier(
+      AzureStorageAccessService storageAccessService, AuthenticatedUserRequest userRequest) {
+    this.storageAccessService = storageAccessService;
+    this.userRequest = userRequest;
+  }
+
+  /**
+   * Attempts to copy all blobs from the source container to the destination container. The calling
+   * principal must have READ-level access on the storage container resource.
```

Comment:
READ on source and WRITE on destination?

Reply:
🤔 an interesting question. The user in this context implicitly has WRITE access as they are the owner of the newly cloned workspace.